AsyncDisplayKit/Private/ASPendingStateController.mm

```diff
@@ -0,0 +1,110 @@
+//
+//  ASPendingStateController.m
+//  AsyncDisplayKit
+//
+//  Created by Adlai Holler on 1/7/16.
+//  Copyright © 2016 Facebook. All rights reserved.
+//
+
+#import "ASPendingStateController.h"
+#import "ASThread.h"
+#import "ASWeakSet.h"
+#import "ASAssert.h"
+#import "ASDisplayNodeInternal.h"
+
+@interface ASPendingStateController()
+{
+  ASDN::Mutex _lock;
+
+  struct ASPendingStateControllerFlags {
+    unsigned pendingFlush:1;
+  } _flags;
+}
+
+@property (nonatomic, strong, readonly) ASWeakSet<ASDisplayNode *> *dirtyNodes;
+@end
+
+@implementation ASPendingStateController
+
+#pragma mark Lifecycle & Singleton
+
+- (instancetype)init
+{
+  self = [super init];
+  if (self) {
+    _dirtyNodes = [ASWeakSet new];
+  }
+  return self;
+}
+
++ (ASPendingStateController *)sharedInstance
+{
+  static dispatch_once_t onceToken;
+  static ASPendingStateController *controller;
+  dispatch_once(&onceToken, ^{
+    controller = [ASPendingStateController new];
+  });
+  return controller;
+}
+
+#pragma mark External API
+
+- (void)flush
+{
+  ASDisplayNodeAssertMainThread();
```

Comment:
Should this method exist?  Maybe just put the assert in -flushNow and rename that to flush.